### PR TITLE
Implement ActionInterface for wishlist/shared

### LINKED ISSUE #29537: [Issue] Implement ActionInterface for wishlist/shared
This issue is automatically created based on existing pull request: magento/magento2#27494: Implement ActionInterface for wishlist/shared

---------

### Preconditions (*)

This is a part of the process of introducing implementations of `ActionInterface`.

### Fixed Issues (if relevant)

N/A

### Expected Result

N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by dxx004**
### Description (*)

This is a part of the process of introducing implementations of `ActionInterface`.

### Fixed Issues (if relevant)

N/A

### Manual testing scenarios (*)

N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29537: Implement ActionInterface for wishlist/shared

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @dxx004,
Could you review and fix static test failures?
Unit test failure seems to me not related to your changes

**Comment by ihor-sviziev** [comment]
Hi @dxx004,
Will you be able to update your PR with requested changes?

**Comment by ihor-sviziev** [comment]
@dxx004  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by lbajsarowicz** [comment]
@ihor-sviziev Reopening due to updates.

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by ihor-sviziev** [comment]
Approved and moved to on hold, waiting for architecture decision

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7557](https://jira.corp.magento.com/browse/ENGCOM-7557) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
FYI this PR was moved to "oh hold" due to internal discussions, reason the same as there https://github.com/magento/magento2/pull/27538#issuecomment-624891873

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @dxx004 ,
We finally have final decision, it clearly described here https://github.com/magento/magento2/pull/27296#pullrequestreview-418541438.

Could you update your PR by bringing back extending abstract action?

**Comment by ihor-sviziev** [comment]
@dxx004,
Will you be able to update your PR?

**Comment by ihor-sviziev** [comment]
@dxx004  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by lbajsarowicz** [comment]
As per @dxx004 request, I'm reopening the PR.
What you need to do is to revert `extends` on `Action` class that was inherited previously.
Additionally, Magento Architects decided to keep the backwards-compatible Constructor, so make sure that the elements in Constructor after rebuilding are in the correct order.

https://github.com/dxx004/magento2/blob/f4744395fa245230386c65d00405dfba665dc868/app/code/Magento/Wishlist/Controller/Shared/Allcart.php#L46-L56

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by dxx004** [comment]
@ihor-sviziev I see some failed tests, but cannot see why that failed. Could you check pls?

**Comment by ihor-sviziev** [comment]
@lenaorobei @slavvka @naydav could you help to identify why builds fail without any report?

**Comment by lenaorobei** [comment]
Here is log message:
```
10:13:21 [Build Functional CE] Repositories code generation... 1/8 [==>-----------------]  12% < 1 sec 46.0 MiBPHP Fatal error:  Access level to Magento\Wishlist\Controller\Shared\Allcart::$resultFactory must be protected (as in class Magento\Framework\App\Action\Action) or weaker in /var/www/html/app/code/Magento/Wishlist/Controller/Shared/Allcart.php on line 20 [] []
```

**Comment by dxx004** [comment]
@magento run all tests

**Comment by dxx004** [comment]
@magento run all tests

**Comment by dxx004** [comment]
@magento run all tests

**Comment by lbajsarowicz** [comment]
@magento run Functional Tests B2B, Functional Tests EE

File: app/code/Magento/Wishlist/Controller/Shared/Allcart.php
Comment (ihor-sviziev) [inline app/code/Magento/Wishlist/Controller/Shared/Allcart.php]: Could you revert all not needed changes, changing property visibility to private, changing properties order?

File: app/code/Magento/Wishlist/Controller/Shared/Allcart.php
Comment (ihor-sviziev) [inline app/code/Magento/Wishlist/Controller/Shared/Allcart.php]: Please revert this change
```suggestion
        WishlistProvider $wishlistProvider,
        ItemCarrier $itemCarrier
```

File: app/code/Magento/Wishlist/Controller/Shared/Allcart.php
Comment (ihor-sviziev) [inline app/code/Magento/Wishlist/Controller/Shared/Allcart.php]: Please revert this change
```suggestion
        $this->wishlistProvider = $wishlistProvider;
        $this->itemCarrier = $itemCarrier;
```

File: app/code/Magento/Wishlist/Controller/Shared/Cart.php
Comment (ihor-sviziev) [inline app/code/Magento/Wishlist/Controller/Shared/Cart.php]: Please revert changing property visibility for all properties.

File: app/code/Magento/Wishlist/Test/Unit/Controller/Shared/AllcartTest.php
Comment (ihor-sviziev) [inline app/code/Magento/Wishlist/Test/Unit/Controller/Shared/AllcartTest.php]: Please revert changing property position, also would be good to make all property private **in tests**

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Comment by engcom-Charlie** [comment]
@magento run Unit Tests, Static Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @dxx004 . During testing, we faced the issue.

**Manual testing scenario:**

1. Login to a customer account and add a product to wishlist.
2. Go to "wishlist" and click "Share wishlist".
3. Enter email, message and send.
4. Logout from customer account.
5. Open a shared wishlist email, Click on "View Product" from email.
6. It landed on the shared wishlist products;
 Click the "Add to Cart" button from any product.

**Actual Result:** :heavy_multiplication_x:  404 error message

Peek 2020-08-11 13-16

The problem is that it has been implemented `HttpGetActionInterface` instead `HttpPostActionInterface` in `app/code/Magento/Wishlist/Controller/Shared/Allcart.php`

With `HttpPostActionInterface` everything works as expected.
Peek 2020-08-11 13-28

@ihor-sviziev @dxx004 Could you take a look?

Thanks!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Charlie could you review and fix failing tests?

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@ihor-sviziev could you please review this PR?

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7779](https://jira.corp.magento.com/browse/ENGCOM-7779) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Login to a customer account and add a product to wishlist;
2. Go to "wishlist" and click "Share wishlist";
3. Enter email, message and send;
4. Logout from the customer account;
5. Open a shared wishlist email, Click on "View Product" from email;
6. It landed on the shared wishlist products;
 Click on the "Add to Cart" button from any product.

**Actual Result:** :heavy_check_mark: The product was added to cart

Screenshot from 2020-08-13 11-02-00

 Click on the "Add All to Cart" button

**Actual Result:** :heavy_check_mark: The products were added to cart

Screenshot from 2020-08-13 11-01-24

**Comment by engcom-Alfa** [comment]
@magento create issue

**Pull request merged**
